lib/okta_oidc.go

```diff
@@ -0,0 +1,249 @@
+		))
+		if err != nil {
+			return err
+		}
+
+		select {
+		case creds := <-credChan:
+			newCreds = *creds
+		case <-ctx.Done():
+			log.Debugf("Didn't get credentials with OIDC. Context done: %s", ctx.Err())
+			return errors.New("Failed to get credentials with OIDC")
+		}
+
+		return nil
+	})
+	return newCreds, err
+}
+
+func (o *OktaOIDCClient) setRedirectHandler(
+	ctx context.Context,
+	cancel context.CancelFunc,
+	credChan chan *sts.Credentials,
+	config oauth2.Config,
+	verifier *oidc.IDTokenVerifier,
+	profileARN string,
+	duration time.Duration,
+) {
+	http.HandleFunc(RedirectPath, func(w http.ResponseWriter, r *http.Request) {
+		// Whatever the result, the authentication is done after we finish
+		// handling this redirected request.
+		defer cancel()
+		// As this is still mainly a CLI tool, we can return this generic
+		// response to the browser on any outcome.
+		defer w.Write([]byte("This tab/window can safely be closed"))
+
+		log.Debug("Received redirect request")
+
+		if r.URL.Query().Get("state") != o.State {
+			log.Error("OIDC redirect state did not match")
+			return
+		}
+
+		oauth2Token, err := config.Exchange(ctx, r.URL.Query().Get("code"),
+			oauth2.SetAuthURLParam("code_verifier", o.CodeVerifier),
+		)
+		if err != nil {
+			log.Errorf("Failed to exchange OIDC token: %s", err.Error())
+			return
+		}
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			log.Error("No id_token field in oauth2 token")
+			return
+		}
+		_, err = verifier.Verify(ctx, rawIDToken)
+		if err != nil {
+			log.Errorf("Failed to verify OIDC token: %s", err.Error())
+			return
+		}
+
+		log.Debug("Verified OIDC token. Using the ID token to assume an AWS role.")
+		sess := session.Must(session.NewSession())
+		svc := sts.New(sess)
+		params := &sts.AssumeRoleWithWebIdentityInput{
+			RoleArn:          aws.String(profileARN),
+			RoleSessionName:  aws.String(o.Username),
+			WebIdentityToken: aws.String(rawIDToken),
+			DurationSeconds:  aws.Int64(int64(duration.Seconds())),
+		}
+		assumeResp, err := svc.AssumeRoleWithWebIdentity(params)
+		if err != nil {
+			log.WithField("role", profileARN).Errorf("Error assuming role with OIDC: %s", err.Error())
+			return
+		}
+		credChan <- assumeResp.Credentials
+	})
+}
+
+func (o *OktaOIDCClient) startRedirectServer() *http.Server {
+	srv := &http.Server{Addr: RedirectAddr}
+	go func() {
+		log.Debugf("Listening for redirect on http://%s/", RedirectAddr)
+		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Debugf("Httpserver: ListenAndServe() error: %s", err)
+		}
+	}()
+	return srv
+}
+
+func (o *OktaOIDCClient) stopRedirectServer(srv *http.Server) {
+	log.Debug("Shutting server down")
+	if err := srv.Close(); err != nil {
+		// This should be safe to ignore in a CLI program, which shuts down
+		// shortly after anyway
+		log.Debug("Failed to shut down server. Ignoring")
+	}
+}
+
+func restoreWindowFocusAfter(f func() error) error {
```

Comment:
How essential is this for users? Running Applescript feels gross and requires the user to acknowledge a security prompt. I'd love to cut this out.